k8s/upgrade/src/preflight_validations.rs

```diff
@@ -1,26 +1,101 @@
-use crate::user_prompt;
-use operators::upgrade::{common::error, controller::reconciler};
+use crate::{
+    constant::SINGLE_REPLICA_VOLUME, upgrade_resources::upgrade::get_pvc_from_uuid, user_prompt,
+};
+use operators::upgrade::common::error;
+use plugin::rest_wrapper::RestClient;
+use std::collections::HashSet;
+use tracing::error;
 
 /// Validation to be done before applying upgrade.
 pub async fn preflight_check() -> Result<(), error::Error> {
     console_logger::info(user_prompt::UPGRADE_WARNING);
-    rebuild_in_progress_validation().await?;
+    if rebuild_in_progress_validation().await? {
+        console_logger::warn(user_prompt::REBUILD_WARNING, "");
+    }
+    single_volume_replica_validation().await?;
     Ok(())
 }
 
-/// Check for rebuild in progress.
-pub async fn rebuild_in_progress_validation() -> Result<(), error::Error> {
-    match reconciler::is_rebuilding().await {
-        Ok(is_rebuilding) => {
-            if is_rebuilding {
-                console_logger::warn(user_prompt::REBUILD_WARNING, "");
-                std::process::exit(1);
-            }
+/// Check single replica volumes.
+pub async fn single_volume_replica_validation() -> Result<(), error::Error> {
+    let mut single_replica_volumes = Vec::new();
+    // The number of volumes to get per request.
+    let max_entries = 200;
+    let mut starting_token = Some(0_isize);
+    let mut volumes = Vec::with_capacity(max_entries as usize);
+
+    // The last paginated request will set the `starting_token` to `None`.
+    while starting_token.is_some() {
+        let vols = RestClient::client()
+            .volumes_api()
+            .get_volumes(max_entries, None, starting_token)
+            .await
+            .map_err(|error| {
+                error!(?error, "Failed to list volumes");
+                error
+            })?;
+
+        let v = vols.into_body();
+        volumes.extend(v.entries);
```

Comment:
nit:- Instead of appending all the volumes in the vector you can simply filter the single volume replicas here and append in the list. That would save the below loop.

```
while let Some(starting_token) = starting_token {
    let vols = RestClient::client()
        .volumes_api()
        .get_volumes(max_entries, None, starting_token)
        .await
        .map_err(|error| {
            error!(?error, "Failed to list volumes");
            error
        })?;

    let single_rep_vol_ids = vols
        .into_body()
        .entries
        .into_iter()
        .filter(|volume| volume.spec.num_replicas == SINGLE_REPLICA_VOLUME)
        .map(|volume| volume.spec.uuid.to_string())
        .collect();

    volumes.extend(single_replica_volumes);

    starting_token = vols.into_body().next_token;
}
```

Something like this probably. Then you can directly create the hashset from the volume vector.